# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: editor/libeditor/HTMLEditor.cpp
# Commit: cdcef4f3f635
# Full Hash: cdcef4f3f63553a28a1055e6f4f051778ca8fb0a
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-09-22 09:56:47
# Regressor Bug: 1789967
# File Overlap Count: 1
# Description:
#   Bug 1789967 - part 2: Make `TextEditor` and `HTMLEditor` implement `EditorBase::CollapseSelectionToEndOfLastLeafNode` by themselves r=m_kato
#   
#   It does different thing for `TextEditor` and `HTMLEditor`, and used almost
#   internally.  Therefore, it should be implemented in the sub classes and
#   we should name them better.
# ==============================================================================

diff -r cc2e8b7c548e -r cdcef4f3f635 editor/libeditor/HTMLEditor.cpp
--- a/editor/libeditor/HTMLEditor.cpp	Thu Sep 22 06:06:54 2022 +0000
+++ b/editor/libeditor/HTMLEditor.cpp	Thu Sep 22 06:17:36 2022 +0000
@@ -790,6 +790,46 @@
   return rv;
 }
 
+NS_IMETHODIMP HTMLEditor::EndOfDocument() {
+  AutoEditActionDataSetter editActionData(*this, EditAction::eNotEditing);
+  if (NS_WARN_IF(!editActionData.CanHandle())) {
+    return NS_ERROR_NOT_INITIALIZED;
+  }
+  nsresult rv = CollapseSelectionToEndOfLastLeafNodeOfDocument();
+  NS_WARNING_ASSERTION(
+      NS_SUCCEEDED(rv),
+      "HTMLEditor::CollapseSelectionToEndOfLastLeafNodeOfDocument() failed");
+  // This is low level API for embedders and chrome script so that we can return
+  // raw error code here.
+  return rv;
+}
+
+nsresult HTMLEditor::CollapseSelectionToEndOfLastLeafNodeOfDocument() const {
+  MOZ_ASSERT(IsEditActionDataAvailable());
+
+  RefPtr<Element> bodyOrDocumentElement = GetRoot();
+  if (NS_WARN_IF(!bodyOrDocumentElement)) {
+    return NS_ERROR_NULL_POINTER;
+  }
+
+  auto pointToPutCaret = [&]() -> EditorRawDOMPoint {
+    nsCOMPtr<nsIContent> lastLeafContent = HTMLEditUtils::GetLastLeafContent(
+        *bodyOrDocumentElement, {LeafNodeType::OnlyLeafNode});
+    if (!lastLeafContent) {
+      return EditorRawDOMPoint::AtEndOf(*bodyOrDocumentElement);
+    }
+    // TODO: We should put caret into text node if it's visible.
+    return lastLeafContent->IsText() ||
+                   HTMLEditUtils::IsContainerNode(*lastLeafContent)
+               ? EditorRawDOMPoint::AtEndOf(*lastLeafContent)
+               : EditorRawDOMPoint(lastLeafContent);
+  }();
+  nsresult rv = CollapseSelectionTo(pointToPutCaret);
+  NS_WARNING_ASSERTION(NS_SUCCEEDED(rv),
+                       "EditorBase::CollapseSelectionTo() failed");
+  return rv;
+}
+
 void HTMLEditor::InitializeSelectionAncestorLimit(
     nsIContent& aAncestorLimit) const {
   MOZ_ASSERT(IsEditActionDataAvailable());